Support for the IS31FL3236 LED Driver as an RGB Matrix Driver

Electro707 opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description

Adding support for the IS31FL3236 LED driver as an RGB matrix driver.

Tested with a prototype keyboard if mine with the IS31FL3236A IC (backwards compatible with the IS31FL3236 IC with the exception of a selectable PWM frequency which isn't used in this driver yet (See TODO)).

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [?] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [See TODO] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

## TODO:
- [ ] The IC supports adjustable current (from IMAX to IMAX/4). As of right now the driver drives the LEDs at IMAX. 
  -  Should we have a callable function to allow that to be changed, or is unnecessary at the moment (would be trivial to add)
- [ ] As IS31FL3236 and the IS31FL3236A are similar, should I add another define for the A version but have it use the same driver (and perhaps add the adjustable PWM for the A version as a callable function only when the A version is used)
- [x] __BUG__: On my device, after 10-20 minutes the LED driver shuts down (so all LEDs turns off) and no LED binded key-press awakes it (but the keyboard still works). Unsure if that's a QMK shutdown feature or a bug in the driver. Need to investigate further.
  - __FIXED__: I had the shutdown pin wired to the microcontroller, and I didn't set that pin to HIGH and left it floating.

fauxpark commented on common_features.mk:
```diff
@@ -286,6 +286,13 @@ endif
         QUANTUM_LIB_SRC += i2c_master.c
     endif
 
+    ifeq ($(strip $(RGB_MATRIX_DRIVER)), IS31FL3236)
+        OPT_DEFS += -DIS31FL3236 -DSTM32_I2C -DHAL_USE_I2C=TRUE
+        COMMON_VPATH += $(DRIVER_PATH)/issi
+        SRC += is31fl3236.c
+        QUANTUM_LIB_SRC += i2c_master.c
+    endif
+
```
Move this before the 3731 stanza.

Electro707 replied:
Should be moved by recent commit

fauxpark commented on docs/feature_rgb_matrix.md:
```diff
@@ -108,7 +108,37 @@ const is31_led g_is31_leds[DRIVER_LED_TOTAL] = {
 Where `X_Y` is the location of the LED in the matrix defined by [the datasheet](https://www.issi.com/WW/pdf/31FL3733.pdf) and the header file `drivers/issi/is31fl3733.h`. The `driver` is the index of the driver you defined in your `config.h` (Only `0` right now).
 
 ---
+# IS31FL3236 :id=is31fl3236
```
As above.

Electro707 replied:
Same as above

fauxpark commented on quantum/rgb_matrix.h:
```diff
@@ -33,6 +33,8 @@
 #    include "is31fl3737.h"
 #elif defined(IS31FL3741)
 #    include "is31fl3741.h"
+#elif defined(IS31FL3236)
```
etc.

Electro707 replied:
Same as above^2

Electro707 commented:
This PR should be ready to merge, all of the review's comments have been addressed last time I checked.

Electro707 commented:
@drashna @fauxpark Do you want to re-open this PR? I know it's been a while, but I do eventually want to merge this.
I will look into re-basing master on this branch to see if there are any new changes to the LED drive that this PR conflicts with since this PR was created.

tzarc commented:
> I will look into re-basing master on this branch 
Just an FYI, core changes need to go to `develop`, not `master`.

Electro707 commented:
Thanks for also changing this PR's merging branch. I think the develop branch is something relatively new, at least since I originally created this PR.

Electro707 commented:
Fixed some issues that occurred when re-basing. Tested with my keypad.
Fixed docs and added the `g_is31_leds` LED definition instead of amusing a sequential connection.
This PR should be all set, at least from my end

drashna commented on drivers/led/issi/is31fl3236.c:
```diff
@@ -0,0 +1,135 @@
+    is31_led led;
+    memcpy_P(&led, (&g_is31_leds[index]), sizeof(led));
+
+    g_pwm_buffer[led.r]  = red;
+    g_pwm_buffer[led.g]  = green;
+    g_pwm_buffer[led.b]  = blue;
+    g_pwm_buffer_update_required = true;
```
```suggestion
    g_pwm_buffer[led.r]          = red;
    g_pwm_buffer[led.g]          = green;
    g_pwm_buffer[led.b]          = blue;
    g_pwm_buffer_update_required = true;
```

Electro707 replied:
Should be fixed in 133d473

drashna commented on drivers/led/issi/is31fl3236.c:
```diff
@@ -0,0 +1,135 @@
+    g_led_control_registers[led.r] = red;
+    g_led_control_registers[led.g] = green;
+    g_led_control_registers[led.b] = blue;
+    g_led_control_update_required = true;
```
```suggestion
    g_led_control_update_required  = true;
```

Electro707 replied:
Should be fixed in 133d473

drashna reviewed (changes requested):
Lint no happy.

Electro707 commented:
@drashna @fauxpark Another request to re-open this PR to merge it, as the review request didn't happen last time
